inference/server/oasst_inference_server/routes/plugins.py

```diff
@@ -0,0 +1,150 @@
+import json
+import secrets
+
+import aiohttp
+import fastapi
+import sqlmodel
+from cryptography.hazmat.primitives import hashes
+from cryptography.hazmat.primitives.kdf.hkdf import HKDF
+from jose import jwe
+from oasst_inference_server import database, deps, models, plugin_utils
+from oasst_inference_server.schemas import plugin as plugin_schema
+from oasst_inference_server.settings import settings
+
+router = fastapi.APIRouter(
+    prefix="/plugins",
+    tags=["plugins"],
+)
+
+# TODO: outsource logic in this file to a PluginRepository or something
+
+
+def derive_plugin_key() -> bytes:
+    """Derive a key from the plugin auth secret."""
+    hkdf = HKDF(
+        algorithm=hashes.SHA256(),
+        length=settings.plugin_auth_length,
+        salt=settings.plugin_auth_salt,
+        info=settings.plugin_auth_info,
+    )
+    key = hkdf.derive(settings.plugin_auth_secret)
+    return key
+
+
+def encrypt_secret(secret: str) -> str:
+    payload = json.dumps({"secret": secret}).encode()
+    key = derive_plugin_key()
+    encrypted: bytes = jwe.encrypt(payload, key)
+    return encrypted.decode()
+
+
+def decrypt_secret(encrypted_secret: str) -> str:
+    key = derive_plugin_key()
+    decrypted = jwe.decrypt(encrypted_secret, key)
+    payload: dict = json.loads(decrypted.decode())
+    secret = payload["secret"]
+    return secret
+
+
+async def get_provider(session: database.AsyncSession, provider: str) -> models.DbPluginOAuthProvider:
+    qry = sqlmodel.select(models.DbPluginOAuthProvider).filter(
+        models.DbPluginOAuthProvider.provider == provider,
+    )
+    db_provider: models.DbPluginOAuthProvider = (await session.exec(qry)).one_or_none()
+    if not db_provider:
+        raise fastapi.HTTPException(status_code=404)
+    return db_provider
+
+
+def get_redirect_uri(provider: str) -> str:
+    return f"{settings.api_root}/plugins/oauth_provider/{provider}/callback"
```

Comment:
I think here we actually need to redirect to the equivalent website URL, not the API